### DISCUSSION TITLE
db diff direction

### DISCUSSION THREAD

**Discussion by dsl400** [supabase/supabase#21054] [Questions]
Our project heavily relies on database functions and in order to streamline the development we decided to store every function definition in a separate file.
We execute the create or update function statement after every modification using a vscode extension.  This approach allowed us to have version control over the function definitions.

The problem we have is generating diff files between the migration folder and our self hosted supabase instance. 

When we run supabase `db diff --db-url .....` the generated sql script contains statemens that revert our local changes.
Instead of a script that is meant to update the remote db  we have a script that remove features from our local db.

Is there any method to configure the direction of the diff tool ?

**Comment by mwoss** [comment]
As far as I know `supabase db diff` always compares schema differences between the target database and a shadow database, and it does not provide the functionality to do it the other way around. You would need to write your own script for that. 
Btw do I understand correctly that you use this command for applying migrations? Why not just `supabase migration`?

**Comment by dsl400** [comment]
Migrations work great when the modifications consist only in adding tables or columns however our project uses many database functions that need incremental modifications.
Spreading db function modifications across all the migrations will make version control useless. 

However, the solution is almost ready. We will use a docker container to create  a shadow database and migra  to extract the diff file

```
docker run --name tmp_postgres -e POSTGRES_PASSWORD=postgres -e POSTGRES_DB=postgres -p 54333:5432 -d supabase/postgres:15.1.0.147 > /dev/null || exit 1
sleep 10
for file in supabase/migrations/*.sql
do
  echo $file
  cat "$file" | docker exec -i tmp_postgres psql -U postgres -d postgres > /dev/null
done
TMP_STRING="postgresql://postgres:postgres@127.0.0.1:54333/postgres"
migra $PG_STRING $TMP_STRING --schema public
docker rm --force tmp_postgres
```

**Reply by mwoss** [reply]
Oh, so you are keeping the latest version of each procedure/function in a separate file and relying on version control to keep it in sync? Interesting approach. Doesn't it complicate DB rollbacks?

**Comment by dsl400** [comment]
It should not. If the diff tool works in both directions you can easily create a diff.sql and apply it.

**Comment by oddanderson** [comment]
I am running into this too

**Comment by attrove-wolfe** [comment]
Interested in getting an answer to this as well.
